src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -198,13 +198,30 @@ private async Task UpdateSymbolTreeInfoAsync(Project project, CancellationToken
                 Debug.Assert(SupportAnalysis(project));
 
                 // Produce the indices for the source and metadata symbols in parallel.
-                var tasks = new List<Task>
+                var isRemoteWorkspace = project.Solution.Workspace.Kind == WorkspaceKind.RemoteWorkspace;
+                Task[] tasks = isRemoteWorkspace
+                    ? UpdateSymbolTreeInfoInServiceProcessAsync(project, cancellationToken)
+                    : UpdateSymbolTreeInfoInCurrentProcessAsync(project, cancellationToken);
+
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+            }
+
+            private Task[] UpdateSymbolTreeInfoInCurrentProcessAsync(Project project, CancellationToken cancellationToken)
+            {
+                return new Task[]
                 {
-                    GetTask(project, () => UpdateSourceSymbolTreeInfoAsync(project, cancellationToken), cancellationToken),
-                    GetTask(project, () => UpdateReferencesAync(project, cancellationToken), cancellationToken)
+                    UpdateSourceSymbolTreeInfoAsync(project, cancellationToken),
+                    UpdateReferencesAsync(project, cancellationToken),
                 };
+            }
 
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+            private Task[] UpdateSymbolTreeInfoInServiceProcessAsync(Project project, CancellationToken cancellationToken)
```

Comment:
:bulb: Use a value tuple here to avoid the array allocation

Reply:
Then how would we await all of them at once?

Reply:
We don't need to await them all at once. Just await one, then the other. 😄

Reply:
Okay, that would work too! 😁

Reply:
note: in the external process we do like spinning them all off at once.  It's good core utilization.

Reply:
> note: in the external process we do like spinning them all off at once

Right:

1. Start first one
2. Start second one
3. Await first one
4. Await second one

Reply:
@sharwell I think @jnm2 pointd out there's a risk in taking that approach.  namely say that first-task cancels/throws and second-task also throws.  If you 'await' the first one, you can end up not observing the second.  Is that a concern?

Reply:
We can handle that with continuations with OnlyOnFaulted, so we handle both?

Reply:
@sharwell Yes, that pattern has bitten me enough times that I strongly recommend Task.WaitAll or else registering all other tasks with some exception handler on failure. It's a big deal when the implementation leaks uncatchable exceptions to the global exception handler on next GC.

Reply:
@jnm2 Unobserved exceptions are no longer an issue starting with .NET 4.5. No effort (I'm aware of) is put into attempting to ensure that faulted tasks are observed.

Reply:
💭 Even if you want to observe both, you can do so with a try/finally. Though I don't see this as an important consideration at this point.

Reply:
@sharwell 

> Unobserved exceptions are no longer an issue starting with .NET 4.5. 

They no longer kill the process, but they have been a huge issue in net45+ for our desktop apps. The .NET Framework (and .NET Core) had a bug in the ADO.NET stack that leaked faulted tasks this way. Even though the SqlExceptions were expected and handled, an implementation detail meant the next GC brought up our unexpected exception restart dialog and there was no way for us to differentiate an actual programming error from an expected SqlException.

> Even if you want to observe both, you can do so with a try/finally.

This has the danger of the `await` in the `finally` replacing the exception from the first await, hiding the first exception when more likely it's the root cause. If you want the caller to wait for both, Task.WhenAll is best. If not, registering an OnlyOnFaulted continuation is best.

Reply:
If this is the logical pattern:

 - Start tasks
 - Wait for task 1
 - Do something with the result of task 1
 - Wait for task 2
 - Do something with the result of task 2

This is the structure that will allow the caller to handle exceptions, rather than bypassing the caller and going straight to the top-level handler (very late, on next GC):

```cs
var task1 = Foo1Async(...);
var task2 = Foo2Async(...);

async Task HandleTask1()
{
    var result1 = await task1;
    // Do something with the result1
}

await Task.WhenAll(HandleTask1(), task2);

var result2 = task2.Result; // (.Result is safe to use because task2.IsCompletedSuccessfully is true)
// Do something with the result2
```

Reply:
Since Task.WhenAll _must_ be used, there is no reason to not create an array by default, since either they'll be allocated via `params` or there's no safe guarantee on the usage. Keeping the array variant.

Reply:
> Since Task.WhenAll must be used

This is not correct. I can show a helper method tomorrow.

Reply:
I'm curious now, too, but I'm happy as long as there's no possibility of leaving a task unobserved.

Reply:
> This has the danger of the await in the finally replacing the exception from the first await, hiding the first exception when more likely it's the root cause.

Then await the tasks in the other order.

- Start first one
- Start second one
- Await second one
- Await first one

> I can show a helper method tomorrow.

If you want to keep both exceptions (which is generally not necessary since ETW traces already capture both in the off change you need them), you can do this:

```
ExceptionDispatchInfo firstException = null;
try
{
    await firstTask;
}
catch (Exception ex)
{
    firstException = ExceptionDispatchInfo.Capture(ex);
}

try
{
    await secondTask;
}
catch (Exception ex) when (firstException != null)
{
    throw new AggregateException(firstException.SourceException, ex);
}
finally
{
    firstException?.Throw();
}
```

Reply:
Hope this is helpful, and I apologize if it confuses things, but I think it's important to not conditionally throw `AggregateException`. `await` was intentionally designed so that the caller can only catch the first (typed) exception without worrying about `catch (AggregateException ex) when (ex.InnerExceptions.Any(...))`. This is another path of pain from what I've seen. (More in depth at `https://github.com/buvinghausen/TaskTupleAwaiter/issues/7`.)

You're exactly right about ETW. My entire concern is making sure that the caller is in control. The method shouldn't in any circumstance go behind the caller's back and schedule a top-level exception for whenever the next GC is.

I didn't say this clearly above. Besides Task.WhenAll, there's also `ContinueWith` with `OnlyOnFaulted` to prevent GC from doing anything. When the second task is awaited later, its exception will be thrown, but if something else throws before it is awaited, its exception is swallowed (as it should be) just like awaiting WhenAll would do.

Reply:
So, do we want to go on the ContinueWith route or how should we proceed?

Reply:
let's step back a bit.  I have a dumb question:  why don't we just pool these arrays? so instead of always allocating at this point, we can just pass in the array we want, fill it, await all of it, then return the array to the pool.

A simple enough solution that satisfies individual concerns around allocations as well as good task hygiene.  Thoughts?

Reply:
Okay, that sounds like a better idea!

Reply:
`#channellingmyinnerjason` (@jasonmalinowski)

Reply:
@CyrusNajmabadi but that would mean allocating a closure for the continuation.

Reply:
So, uh, hello. I'm apparently @CyrusNajmabadi's inner Jason. What's the question? :-)

Reply:
You could be cutesy and just manually do .ContinueWith.  Specifically, the overload that let's you pass your own data along.  This would allow for a non-allocating closure since you wouldn't capture anything, but would just pass the state along you needed (in this case, the array).

Reply:
I've updated it, can you PTAL?